lib/src/layer/tile_layer.dart

```diff
@@ -571,21 +870,199 @@ class _TileLayerState extends State<TileLayer> {
     return newCoords;
   }
 
-  CustomPoint _getTilePos(Coords coords) {
-    var level = _levels[coords.z];
-    return coords.scaleBy(getTileSize()) - level.origin;
+  Bounds _pxBoundsToTileRange(Bounds bounds) {
+    var tileSize = getTileSize();
+    return Bounds(
+      bounds.min.unscaleBy(tileSize).floor(),
+      bounds.max.unscaleBy(tileSize).ceil() - const CustomPoint(1, 1),
+    );
+  }
+
+  bool _noTilesToLoad() {
+    for (var entry in _tiles.entries) {
+      if (entry.value.loaded == null) {
+        return false;
+      }
+    }
+    return true;
   }
 }
 
-class Tile {
-  final Coords coords;
+typedef void TileReady(Coords<double> coords, dynamic error, Tile tile);
+
+class Tile implements Comparable<Tile> {
+  final String coordsKey;
+  final Coords<double> coords;
+  final CustomPoint<num> tilePos;
+  final ImageProvider imageProvider;
+  final Level level;
+
   bool current;
+  bool retain;
+  bool active;
+  bool loadError;
+  DateTime loaded;
+
+  AnimationController animationController;
+  double get opacity => animationController == null
+      ? (active ? 1.0 : 0.0)
+      : animationController.value;
+
+  // callback when tile is ready / error occurred
+  // it maybe be null forinstance when download aborted
+  TileReady tileReady;
+  ImageInfo imageInfo;
+  ImageStream _imageStream;
+  ImageStreamListener _listener;
+
+  Tile({
+    this.coordsKey,
+    this.coords,
+    this.tilePos,
+    this.imageProvider,
+    this.tileReady,
+    this.level,
+    this.current = false,
+    this.active = false,
+    this.retain = false,
+    this.loadError = false,
+  }) {
+    try {
```

Comment:
@maRci002 - I think this constructor body introduced error reported in #608. The reason why PR https://github.com/johnpryan/flutter_map/pull/609 works is that it replaces the state in the widget-tree, forcing a fresh _TileLayerState to be constructed. 


In it self, the code is not wrong, the problem occurs in `_addTile`, where assignment to array `_tiles[tileCoordsToKey]` is done after the body is executed. 
```
_tiles[tileCoordsToKey] = Tile(
      coords: coords,
      coordsKey: tileCoordsToKey,
      tilePos: _getTilePos(coords),
      current: true,
      level: _levels[coords.z],
      imageProvider: options.tileProvider.getImage(_wrapCoords(coords), options),
      tileReady: _tileReady,
    );
```
The callback method `tileReady` is called indirectly from the method  `_tileOnLoad` when the image has been loaded. If the tile is already cached, the method will return immediately, invoking `tileReady` before the assignment to `_tiles[tileCoordsToKey]` is completed. In the actual callback method `_tileReady`, [the assignment is evaluated](https://github.com/johnpryan/flutter_map/pull/572/files#diff-d0fc1fe821e34169c8068a76a40415f3R826).
```
var key = _tileCoordsToKey(coords);
    tile = _tiles[key];
    if (null == tile) {
      return;
    }
```  
and as you can see, the method will return before it is rendered to the canvas. 

**Working solution**

- Removing the code from constructor body
- using the new method in `loadTileImage` in PR https://github.com/johnpryan/flutter_map/pull/584/files#diff-d0fc1fe821e34169c8068a76a40415f3R1007 to load image expliclity in `_addTile`

I've done this change in https://github.com/DISCOOS/flutter_map/blob/track_change_and_evict_errors/lib/src/layer/tile_layer.dart#L887 and it solves the problem:
```
void _addTile(Coords<double> coords) {
    var tileCoordsToKey = _tileCoordsToKey(coords);
    _tiles[tileCoordsToKey] = Tile(
      coords: coords,
      coordsKey: tileCoordsToKey,
      tilePos: _getTilePos(coords),
      current: true,
      level: _levels[coords.z],
      imageProvider: options.tileProvider.getImage(_wrapCoords(coords), options),
      tileReady: _tileReady,
    );
    _tiles[tileCoordsToKey].loadTileImage();
  }
```

I could make an PR with the change to #584 if you're interested?

Reply:
I'll checkout your suggestions soon. However #609 works because if drawer detects if user wants to navigate to the same menu then it just closes drawer instead of reopen same menu. You can test it if you pan away then if you try to reopen same menu then map current location stays the same as before.

Reply:
Yes, of course. My bad.